OpenSim/Common/DataAdapter.h

```diff
@@ -0,0 +1,225 @@
+    @param[in] source           string identifying the source to be opened
+    @param[in] sourceTypeName   string identifying the source type used to
+                                determine which concrete adapter is used.
+    @return DataTable           the resulting in memory data from the source
+        */
+    AbstractDataTable* createDataTable(const std::string& source,
+        const std::string& sourceTypeName);
+
+    /** Direct DataAdpater data flow into the provided DataTable. */
+//TODO? DataTable& operator>>(DataTable& table);
+
+    /** Direct DataTable data out through the DataAdapter. */
+//TODO? DataAdapter& operator<<(const DataTable& table);
+
+    /** Adapter has access to the data source for reading and/or writing? */
+    bool hasDataAccess() const;
+
+    /** Determine read/write access of the adapter to the data source.*/
+    std::ios_base::openmode getAccessMode() const;
+
+    /** Open the data source to enable access to read and/or writing of 
+        data according to the access mode. If successful returns true.
+        hasDataAccess() will also return true.*/
+    bool openDataSource();
+
+    /** Close the connection used to access the data source. 
+        Returns true if successfully closed access to data. hasDataAccess() 
+        also return false.*/
+    bool closeDataSource();
+
+    /** Give the adapter an opportunity to prepare itself for reading data
+        into a DataTable. Adapter keeps a reference to the DataTable for
+        all subsequent read access.
+    @param[out] table  The DataTable to be filled by the DataAdapter. */
+    void prepareForReading(AbstractDataTable& table);
+
+    /** Give the adapter an opportunity to prepare itself for writing out a
+        DataTable. Adapter keeps a reference to the DataTable for all
+        subsequent write access.
+    @param[in] table  The DataTable to be written out by the DataAdapter. */
+    void prepareForWriting(const AbstractDataTable& table);
+
+    /** Read in rows of the DataTable from this adapter's data source.
+    If a file this can be all the rows of data. If the source is a stream
+    or a device, this may be a row in time or the last few buffered rows.
+    @return  done  returns true if there is no more data to be read in. */
+    bool read();
+
+    /** Write out rows of the DataTableas to this adapter's data sink.
+    If a file this can be all the rows of data. If the sink is a stream
+    or a device, this may the row(s) up to the current time.
+    @return  done  returns true if the table is completely written out. */
+    bool write();
+
+    /** Convenience methods to interpret access mode */
+    /** Return true if DataAdpater expects read access of the data source */
+    bool isReadAccess() const;
+    /** Return true if DataAdpater expects write access on the data source */
+    bool isWriteAccess() const;
+
+protected:
+    /** Only derived DataAdapters can default construct the base */
+    DataAdapter() {};
```

Comment:
Spurious semicolon. You could also write `DataAdapter() = default;` which I've grown to like.